File: server/src/routes/rag.js
// server/src/routes/rag.js
import express                  from 'express';
import { ragChainWithSources }  from '../chains/rag-chain.js';

const router = express.Router();

router.post('/query', async (req, res) => {
  const { question } = req.body;

  if (!question) return res.status(400).json({ error: 'question 不能为空' });

  res.setHeader('Content-Type', 'text/event-stream; charset=utf-8');
  res.setHeader('Cache-Control', 'no-cache');
  res.setHeader('Connection', 'keep-alive');
  res.setHeader('X-Accel-Buffering', 'no');

  const send = (type, data) =>
    res.write(`data: ${JSON.stringify({ type, ...data })}\n\n`);

  try {
    const result = await ragChainWithSources.invoke({ question });

    if (result.sources?.length) {
      send('sources', { sources: result.sources });
    }

    send('answer', { content: result.answer });
    send('done', {});
    res.end();
  } catch (err) {
    console.error('[RAG Error]', err.message);
    send('error', { content: '查询出错，请重试' });
    res.end();
  }
});

export default router;

File: server-py/app/routers/rag.py
import json

from fastapi import APIRouter
from fastapi.responses import JSONResponse, StreamingResponse
from pydantic import BaseModel

from app.chains.rag_chain import rag_chain_with_sources

router = APIRouter()


class RagRequest(BaseModel):
    question: str


def _send(event_type, data):
    return f"data: {json.dumps({'type': event_type, **data}, ensure_ascii=False)}\n\n"


@router.post("/query")
async def rag_query(req: RagRequest):
    if not req.question:
        return JSONResponse(status_code=400, content={"error": "question 不能为空"})

    async def event_generator():
        try:
            result = rag_chain_with_sources.invoke({"question": req.question})

            if result.get("sources"):
                yield _send("sources", {"sources": result["sources"]})

            yield _send("answer", {"content": result["answer"]})
            yield _send("done", {})
        except Exception as err:
            print(f"[RAG Error] {err}")
            yield _send("error", {"content": "查询出错，请重试"})

    return StreamingResponse(event_generator(), media_type="text/event-stream")
